### PR TITLE
Add widget for overmap

### PR DESCRIPTION
**Pull request by wapcaplet**
#### Summary

Features "Add widget for overmap"

#### Purpose of change

Contribute to #53957 

Provide an overmap widget with customizable width and height, that can be added to the custom sidebar.

#### Describe the solution

- [x] Add `display::overmap_tile_symbol_color`, returning the symbol and color at a given overmap tripoint
- [x] Add `display::overmap_note_symbol_color` for the nontrivial task of parsing colors and symbols from a user-defined overmap note
- [x] Add `display::colorized_overmap_text` to return a newline-separated fully-colorized overmap rendering of an arbitrary width and height
- [x] Add `overmap_text` to `widget_var` implementations
- [x] Make it possible to pass `"width": 0` (or simply omit "width") to have overmap fill available width
- [x] Make it possible to "label" the overmap widget (for the purpose of being a standalone sidebar section), but to hide the visible label (without resorting to a layout-wrapper workaround)
- [x] Add widget JSONs for custom sidebar
- [x] Test cases
- [x] Integrate caching patch
- [x] Add `overmap_loc_text` widget to show current overmap coordinates, like `LEVEL 0 1'160, 0'155`
- [x] Get mission marker to render at edge of map when mission target is off-map

#### Describe alternatives you've considered

This widget does not yet cooperate in a "columns" layout side-by-side with another multi-line widget, so there's not much point (yet) in an overmap widget that isn't the full width of the sidebar (using "width": 0).

Sharing horizontal space with other widgets is an essential part of making this overmap widget work with the "classic" sidebar layout, as well as parts of the "labels" layout (where place/weather are shown next to the map). This flexibility in widgets is highly desirable, but I decided it's worth saving for a future PR.

#### Testing

Switch to the "custom" sidebar, and make the "Overmap" visible. Travel around, see the world, watch out for hordes!

#### Additional context

Final version, replacing the previously hard-coded 7-line overmap in the "custom" sidebar with the new widget version (also 7 lines):

image

Players may easily change their overmap height by editing `data/json/ui/overmap.json`, and setting "height" to another value.

Final version, different town, including the new "Location" section:

image

Builds on @dseguin 's multi-line widgets #54050

### CODE REVIEW

**Comment by dseguin** [comment]
This is so cool!

It occurred to me that I never test these things in curses mode, so I tried this in a VT102 emulator:
normal_view

After stretching the width to 32:
normal_view_w32

After increasing the height to 20:
normal_view_w32_h20

Looks great, no issues so far!

**Comment by wapcaplet** [comment]
That's awesome! I have never once run the game in curses mode, but should give it a try as well.

For some reason I couldn't get this map widget to work unless I wrapped it in a "rows" layout. Putting the "overmap_13x7" widget directly into the main sidebar, all the lines ran together. I stared at our newly-revised layout and multiline code for a while but it's getting complex enough to evade my grasp now... Could be I just need to sleep on it, but I would of course welcome any insight you have.

**Comment by dseguin** [comment]
> For some reason I couldn't get this map widget to work unless I wrapped it in a "rows" layout. Putting the "overmap_13x7" widget directly into the main sidebar, all the lines ran together.

That's strange, guess I should have considered that case...
I'll take a look at it. I feel like it might just be a simple mistake on my part.

**Comment by dseguin** [comment]
Well, I was right about it being a simple mistake:
```diff
diff --git a/src/widget.cpp b/src/widget.cpp
index 15bbc6156a..2d9f855b32 100644
--- a/src/widget.cpp
+++ b/src/widget.cpp
@@ -430,7 +430,7 @@ window_panel widget::get_window_panel( const int width, const int req_height )
         for( const widget_id &wid : _widgets ) {
             height += wid->_height > 0 ? wid->_height : 1;
         }
-    } else if( _style == "widget" ) {
+    } else if( _style == "widget" || _style == "text" ) {
         height = _height > 1 ? _height : req_height;
     }
     // Minimap and log do not have a predetermined height
```

That simple fix turned this:
overmap_widget_notfixed

... into this:
overmap_widget_fix

**Comment by wapcaplet** [comment]
I included your patch, but I'm still using the layout wrapper around the overmap widget because we don't currently have a way to hide a widget's label. Adding `overmap_13x7` directly to the sidebar either won't have a label, or it has a label that interferes with the map drawing.

I got the mission marker working when the target is currently visible in the overmap area, but the off-map indicator (red arrow at the edge of the map) is a separate blob of annoyingly specialized code (I should know, I wrote it) that isn't really compatible with string-based widget drawing.

I haven't enabled the horde indicators yet either; the existing code leaves a 1-omt space free of `Zz`s around the avatar position, so I just need to refactor it a bit to make it work again.

Anyhow, progress continues. I was worried this widget would be a huge pain to implement, but so far it hasn't been too bad.

**Comment by dseguin** [comment]
> we don't currently have a way to hide a widget's label. Adding `overmap_13x7` directly to the sidebar either won't have a label, or it has a label that interferes with the map drawing.

I think I have an elegant solution in mind that will address some other issues as well. Using flags, we could add special behaviours to widgets:
```json
{
  "id": "overmap_13x7",
  "type": "widget",
  "label": "NU-Overmap",
  "var": "overmap_text",
  "style": "text",
  "width": 13,
  "height": 7,
  "flags": [ "NOLABEL", "PAD_CENTER" ]
}
```
Where `NOLABEL` => don't display the label in the sidebar, and `PAD_CENTER` => add padding to the left so the widget is centered. (plus more flags, possibly)

I can start working on that if it works for you.

**Comment by wapcaplet** [comment]
> I can start working on that if it works for you.

Nice idea! Yeah, go for it.

**Comment by dseguin** [comment]
Turns out that adding flags was super simple #54190 :)

**Comment by wapcaplet** [comment]
I did some more refactoring of the mission arrow indicator - it's still not working for the new widget overmap, but is better modularized for the old one.

There are both `13x7` and `Fullx9` versions for comparison now:

image

Other outstanding problems (highlighted in pink):

- The green horde `Zz` indicators *should* be working, but aren't
- There are two leading spaces in front of the widget on the main sidebar, and an ellipsis at the end

**Comment by dseguin** [comment]
> There are two leading spaces in front of the widget on the main sidebar, and an ellipsis at the end

Here's the the fix for the spacing issue:
```diff
diff --git a/src/widget.cpp b/src/widget.cpp
index dc7eacb7b6..68ed9a2b32 100644
--- a/src/widget.cpp
+++ b/src/widget.cpp
@@ -793,7 +793,7 @@ static std::string append_line( const std::string &line, bool first_row, unsigne
 {
     std::string ret;
     // Width used by label, ": " and value, using utf8_width to ignore color tags
-    unsigned int used_width = utf8_width( line, true );
+    unsigned int used_width = utf8_width( line, true ) + 1;
     if( first_row ) {
         const std::string tlabel = label.translated();
         // If label is empty or omitted, don't reserve space for it
```

It turns out that the actual length used for drawing is one more than what is reported, so correcting that gives the right number of padding spaces:

pad_fix

If you want, you could also use `W_PAD_NONE` or `W_PAD_CENTER` to adjust the alignment:

left_align
center_align

**Comment by wapcaplet** [comment]
> Just a little bit of an explanation about the padding bug

Thanks for the extra explanation, I was a bit puzzled by the +1 fix. Not sure if I'm qualified to say whether newlines should subtract from the length there in `utf8_width` or not... newlines aren't printable, so it kind of makes sense that they would be omitted from the count.

**Comment by wapcaplet** [comment]
So I backed out that +1 patch, since it caused something like 63 test cases to fail due to all the layouts being the wrong total width. Guess that'll take some rethinking.

I'm also hoping to take another crack at the missing horde indicator. I would really like to get the overmap widget to work side-by-side with another multi-line widget, but that too will take some rethinking.

**Comment by dseguin** [comment]
> So I backed out that +1 patch, since it caused something like 63 test cases to fail due to all the layouts being the wrong total width. Guess that'll take some rethinking.

If you're hesitant about the `utf8_width()` solution, you could fix it specifically for this case with something like:
```diff
diff --git a/src/widget.cpp b/src/widget.cpp
index 35c1563919..0f75542655 100644
--- a/src/widget.cpp
+++ b/src/widget.cpp
@@ -784,9 +784,14 @@ std::string widget::graph( int value, int value_max )
 static std::string append_line( const std::string &line, bool first_row, unsigned int max_width,
                                 const translation &label, int align )
 {
+    // Widgets may be empty at times
+    if( line.empty() ) {
+        return "";
+    }
+
     std::string ret;
     // Width used by label, ": " and value, using utf8_width to ignore color tags
-    unsigned int used_width = utf8_width( line, true );
+    unsigned int used_width = utf8_width( line, true ) + line.back() == '\n' ? 1 : 0;
     if( first_row ) {
         const std::string tlabel = label.translated();
         // If label is empty or omitted, don't reserve space for it
```

This leaves single-line widgets alone while fixing multiline widgets. Again, that pesky `utf8_width()` is subtracting 1 for every control character (ie: `\n`), so you'll need to add 1 to fix the padding.

**Comment by wapcaplet** [comment]
Oh, of course, it only needs the extra 1 if there was a newline. Thanks for the fix! I adjusted it a little further for readability in c74ea54.

**Comment by wapcaplet** [comment]
Regarding the horde indicators, I've been scratching my head at this for a while before I noticed that even the *original* horde indicator is behaving strangely.

Here's a save game with some hordes around the character in the city: [HordeMap.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/7858722/HordeMap.zip)

My character is positioned at the top-right corner of an overmap tile. Stepping to the east or north causes some of the horde `Z` indicators to disappear and reappear. Both the original hard-coded overmap (bottom) and the new widget-based overmap (top, larger) have disappearing horde `Z`s, but each of them slightly different. Here I made them pink:

disappearing-hordes

In other words, they both seem to work poorly... I didn't test in master for comparison to see if this is a glitch I introduced, but at least the new ones are working, kind of.

**Comment by wapcaplet** [comment]
> (Hope I'm not spamming you with solutions at this point)

Not at all, I'm okay with this pattern where I find a problem and you fix it for me! Nice detective work again, I will integrate this fix.

**Comment by dseguin** [comment]
The mission indicator is alive!

om_w_mission

<details>
<summary>Here's the diff in case you want to include it:</summary>

```diff
diff --git a/src/display.cpp b/src/display.cpp
index 6720020154..a8bff83428 100644
--- a/src/display.cpp
+++ b/src/display.cpp
@@ -1181,7 +1181,7 @@ std::pair<std::string, nc_color> display::overmap_note_symbol_color( const std::
 }
 
 std::pair<std::string, nc_color> display::overmap_tile_symbol_color( const avatar &u,
-        const tripoint_abs_omt &omt )
+        const tripoint_abs_omt &omt, bool edge_tile )
 {
     std::string ter_sym;
     nc_color ter_color = c_light_gray;
@@ -1216,8 +1216,15 @@ std::pair<std::string, nc_color> display::overmap_tile_symbol_color( const avata
     const tripoint_abs_omt u_loc = u.global_omt_location();
 
     // If there's a valid mission target, and avatar isn't already there, mark it with a red background
-    if( target != overmap::invalid_tripoint && target.xy() == omt.xy() && target.xy() != u_loc.xy() ) {
-        ter_color = red_background( ter_color );
+    if( target != overmap::invalid_tripoint && target.xy() != u_loc.xy() ) {
+        if( target.xy() == omt.xy() ) {
+            ter_color = red_background( ter_color );
+        } else if( edge_tile ) {
+            std::vector<tripoint_abs_omt> plist = line_to( u_loc, target );
+            if( std::find( plist.begin(), plist.end(), omt ) != plist.end() ) {
+                ter_color = red_background( ter_color );
+            }
+        }
     }
 
     // Show hordes on minimap, leaving a one-tile space around the player
@@ -1252,9 +1259,12 @@ std::string display::colorized_overmap_text( const avatar &u, const int width, c
     for( int row = -( height / 2 ); row <= height - ( height / 2 ) - 1; row++ ) {
         // Scan across the width of the row
         for( int col = -( width / 2 ); col <= width - ( width / 2 ) - 1; col++ ) {
+            // Is this the edge of the overmap widget?
+            const bool edge = row == -( height / 2 ) || row == height - ( height / 2 ) - 1 ||
+                              col == -( width / 2 ) || col == width - ( width / 2 ) - 1;
             // Get colorized symbol for this point
             const tripoint_abs_omt omt( center_xyz.xy() + point( col, row ), here.get_abs_sub().z );
-            std::pair<std::string, nc_color> sym_color = display::overmap_tile_symbol_color( u, omt );
+            std::pair<std::string, nc_color> sym_color = display::overmap_tile_symbol_color( u, omt, edge );
 
             // Highlight player character location in the center
             if( row == 0 && col == 0 ) {
diff --git a/src/display.h b/src/display.h
index b42db9603e..3cdbea8022 100644
--- a/src/display.h
+++ b/src/display.h
@@ -139,7 +139,7 @@ std::pair<std::string, nc_color> rad_badge_text_color( const Character &u );
 
 // Colorized symbol for the overmap tile at the given location
 std::pair<std::string, nc_color> overmap_tile_symbol_color( const avatar &u,
-        const tripoint_abs_omt &omt );
+        const tripoint_abs_omt &omt, bool edge_tile );
 // Colorized symbol for an overmap note, given its full text
 std::pair<std::string, nc_color> overmap_note_symbol_color( const std::string note_text );
 // Mission marker position as an offset within an overmap of given width and height
```

</details>

<details>
<summary>Basic explanation:</summary>

1. Check if the omt we're processing is an edge tile
2. Pass the result to `display::overmap_tile_symbol_color`
3. When checking the mission indicator, draw a line from our current position to the mission target. If the current omt is an edge tile and it's within the line, that's our mission indicator.

</details>

**Comment by dseguin** [comment]
Here's an amendment to the previous diff to fix the multi-edge indicator:

```diff
diff --git a/src/display.cpp b/src/display.cpp
index a8bff83428..75a98b243d 100644
--- a/src/display.cpp
+++ b/src/display.cpp
@@ -1181,7 +1181,7 @@ std::pair<std::string, nc_color> display::overmap_note_symbol_color( const std::
 }
 
 std::pair<std::string, nc_color> display::overmap_tile_symbol_color( const avatar &u,
-        const tripoint_abs_omt &omt, bool edge_tile )
+        const tripoint_abs_omt &omt, bool edge_tile, bool &found_mi )
 {
     std::string ter_sym;
     nc_color ter_color = c_light_gray;
@@ -1215,14 +1215,17 @@ std::pair<std::string, nc_color> display::overmap_tile_symbol_color( const avata
     const tripoint_abs_omt target = u.get_active_mission_target();
     const tripoint_abs_omt u_loc = u.global_omt_location();
 
+    found_mi = false;
     // If there's a valid mission target, and avatar isn't already there, mark it with a red background
     if( target != overmap::invalid_tripoint && target.xy() != u_loc.xy() ) {
         if( target.xy() == omt.xy() ) {
             ter_color = red_background( ter_color );
+            found_mi = true;
         } else if( edge_tile ) {
             std::vector<tripoint_abs_omt> plist = line_to( u_loc, target );
             if( std::find( plist.begin(), plist.end(), omt ) != plist.end() ) {
                 ter_color = red_background( ter_color );
+                found_mi = true;
             }
         }
     }
@@ -1255,16 +1258,18 @@ std::string display::colorized_overmap_text( const avatar &u, const int width, c
         return disp_om_cache.get_val();
     }
 
+    bool found_mi = false;
     // Scan each row of overmap tiles
     for( int row = -( height / 2 ); row <= height - ( height / 2 ) - 1; row++ ) {
         // Scan across the width of the row
         for( int col = -( width / 2 ); col <= width - ( width / 2 ) - 1; col++ ) {
             // Is this the edge of the overmap widget?
-            const bool edge = row == -( height / 2 ) || row == height - ( height / 2 ) - 1 ||
-                              col == -( width / 2 ) || col == width - ( width / 2 ) - 1;
+            const bool edge = !found_mi && (
+                              row == -( height / 2 ) || row == height - ( height / 2 ) - 1 ||
+                              col == -( width / 2 ) || col == width - ( width / 2 ) - 1 );
             // Get colorized symbol for this point
             const tripoint_abs_omt omt( center_xyz.xy() + point( col, row ), here.get_abs_sub().z );
-            std::pair<std::string, nc_color> sym_color = display::overmap_tile_symbol_color( u, omt, edge );
+            std::pair<std::string, nc_color> sym_color = display::overmap_tile_symbol_color( u, omt, edge, found_mi );
 
             // Highlight player character location in the center
             if( row == 0 && col == 0 ) {
diff --git a/src/display.h b/src/display.h
index 3cdbea8022..fa753ceb17 100644
--- a/src/display.h
+++ b/src/display.h
@@ -139,7 +139,7 @@ std::pair<std::string, nc_color> rad_badge_text_color( const Character &u );
 
 // Colorized symbol for the overmap tile at the given location
 std::pair<std::string, nc_color> overmap_tile_symbol_color( const avatar &u,
-        const tripoint_abs_omt &omt, bool edge_tile );
+        const tripoint_abs_omt &omt, bool edge_tile, bool &found_mi );
 // Colorized symbol for an overmap note, given its full text
 std::pair<std::string, nc_color> overmap_note_symbol_color( const std::string note_text );
 // Mission marker position as an offset within an overmap of given width and height
```

**Comment by dseguin** [comment]
Final amendment (I hope). This fixes the case where the mission target is on the edge:

<details>
<summary>Diff patch (on top of previous):</summary>

```diff
diff --git a/src/display.cpp b/src/display.cpp
index 75a98b243d..e68df1c10c 100644
--- a/src/display.cpp
+++ b/src/display.cpp
@@ -1259,17 +1259,25 @@ std::string display::colorized_overmap_text( const avatar &u, const int width, c
     }
 
     bool found_mi = false;
+    const int left = -( width / 2 );
+    const int right = width + left - 1;
+    const int top = -( height / 2 );
+    const int bottom = height + top - 1;
     // Scan each row of overmap tiles
-    for( int row = -( height / 2 ); row <= height - ( height / 2 ) - 1; row++ ) {
+    for( int row = top; row <= bottom; row++ ) {
         // Scan across the width of the row
-        for( int col = -( width / 2 ); col <= width - ( width / 2 ) - 1; col++ ) {
+        for( int col = left; col <= right; col++ ) {
             // Is this the edge of the overmap widget?
-            const bool edge = !found_mi && (
-                              row == -( height / 2 ) || row == height - ( height / 2 ) - 1 ||
-                              col == -( width / 2 ) || col == width - ( width / 2 ) - 1 );
+            const bool edge = !found_mi && !(
+                                  mission_xyz.x() >= center_xyz.x() + left &&
+                                  mission_xyz.x() <= center_xyz.x() + right &&
+                                  mission_xyz.y() >= center_xyz.y() + top &&
+                                  mission_xyz.y() <= center_xyz.y() + bottom ) &&
+                              ( row == top || row == bottom || col == left || col == right );
             // Get colorized symbol for this point
             const tripoint_abs_omt omt( center_xyz.xy() + point( col, row ), here.get_abs_sub().z );
-            std::pair<std::string, nc_color> sym_color = display::overmap_tile_symbol_color( u, omt, edge, found_mi );
+            std::pair<std::string, nc_color> sym_color = display::overmap_tile_symbol_color( u, omt, edge,
+                    found_mi );
 
             // Highlight player character location in the center
             if( row == 0 && col == 0 ) {
```

</details>

Demo:

**Comment by wapcaplet** [comment]
Awesome! I think I got the patches integrated correctly now, and it's working great for me.

**Comment by dseguin** [comment]
Ironically, getting the mission indicator working messed up the widget tests :P

```diff
diff --git a/tests/widget_test.cpp b/tests/widget_test.cpp
index c3b0198ed2..e5e922a35a 100644
--- a/tests/widget_test.cpp
+++ b/tests/widget_test.cpp
@@ -978,9 +978,10 @@ TEST_CASE( "multi-line overmap text widget", "[widget][overmap]" )
     SECTION( "field" ) {
         const std::string brown_dot = "<color_c_brown>.</color>";
         const std::string h_brown_dot = "<color_h_brown>.</color>";
+        const std::string brown_dot_red = "<color_c_brown_red>.</color>";
         const std::vector<std::string> field_3x3 = {
             brown_dot, brown_dot, brown_dot, "\n",
-            brown_dot, h_brown_dot, brown_dot, "\n",
+            brown_dot, h_brown_dot, brown_dot_red, "\n",
             brown_dot, brown_dot, brown_dot, "\n"
         };
 
@@ -992,9 +993,10 @@ TEST_CASE( "multi-line overmap text widget", "[widget][overmap]" )
     SECTION( "forest" ) {
         const std::string green_F = "<color_c_green>F</color>";
         const std::string h_green_F = "<color_h_green>F</color>";
+        const std::string green_F_red = "<color_c_green_red>F</color>";
         const std::vector<std::string> forest_3x3 = {
             green_F, green_F, green_F, "\n",
-            green_F, h_green_F, green_F, "\n",
+            green_F, h_green_F, green_F_red, "\n",
             green_F, green_F, green_F, "\n"
         };
 
@@ -1006,9 +1008,10 @@ TEST_CASE( "multi-line overmap text widget", "[widget][overmap]" )
     SECTION( "central lab" ) {
         const std::string blue_L = "<color_c_light_blue>L</color>";
         const std::string h_blue_L = "<color_h_light_blue>L</color>";
+        const std::string blue_L_red = "<color_c_light_blue_red>L</color>";
         const std::vector<std::string> lab_3x3 = {
             blue_L, blue_L, blue_L, "\n",
-            blue_L, h_blue_L, blue_L, "\n",
+            blue_L, h_blue_L, blue_L_red, "\n",
             blue_L, blue_L, blue_L, "\n"
         };
 
```

**Comment by wapcaplet** [comment]
If by "messed up" you mean "the mission marker is now tested too", that's fine with me!

I saw you had added those to do cache-busting but it's a nice side-effect that it actually works.

**Comment by wapcaplet** [comment]
The old map used red asterisks for off-map pointers, which I kind of like, so I restored that behavior - keeping the red background version when the mission target is enclosed within the map:

asterisk-offscreen

Just another rebase to clean things up, and I think this will finally be ready.
